src/acp/rule.test.ts

```diff
@@ -0,0 +1,704 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { jest, describe, it, expect } from "@jest/globals";
+import { asIri, createThing } from "../thing/thing";
+import {
+  addForbiddenRuleToPolicy,
+  addOptionalRuleToPolicy,
+  addRequiredRuleToPolicy,
+  getForbiddenRuleOnPolicyAll,
+  getOptionalRuleOnPolicyAll,
+  getRequiredRuleOnPolicyAll,
+  Rule,
+  setForbiddenRuleOnPolicy,
+  setOptionalRuleOnPolicy,
+  setRequiredRuleOnPolicy,
+} from "./rule";
+import { DataFactory } from "n3";
+import { Thing, ThingPersisted } from "../interfaces";
+import { Policy } from "./policy";
+
+const ACP_ANY = "http://www.w3.org/ns/solid/acp#anyOf";
+const ACP_ALL = "http://www.w3.org/ns/solid/acp#allOf";
+const ACP_NONE = "http://www.w3.org/ns/solid/acp#noneOf";
+
+const mockRule = (url: string) =>
+  createThing({
+    url,
+  });
+
+const addAll = (
+  thing: ThingPersisted,
+  predicate: string,
+  objects: ThingPersisted[]
+): void => {
+  objects.forEach((objectToAdd: ThingPersisted) => {
+    thing.add(
+      DataFactory.quad(
+        DataFactory.namedNode(asIri(thing)),
+        DataFactory.namedNode(predicate),
+        DataFactory.namedNode(asIri(objectToAdd))
+      )
+    );
+  });
+};
+
+const mockPolicy = (
```

Comment:
This doesn't seem to be an actual 'mock' at all - it's creating an actual Policy dataset. I'd just name it `createPolicy()` (and it seems like a handy little helper function that might be useful in the main source, and not just in tests...?).

Reply:
It's creating an actual Policy with fake data in it - sounds like a mock to me?

(It also mutates its input, which we shouldn't be doing in the main source.)

Reply:
Well, I certainly always think of a mock as mocking behaviour (i.e. "**_So in simple words, mocking is creating objects that simulate the behavior of real objects._**"). This is just test data in a perfectly valid, un-mocked Thing object. Personally I'd just rename this to `createTestPolicy()`. (Likewise, if I had a test WebID value, I wouldn't call that `const mockWebId = new Iri('https://...);`, I'd just call it a `testWebId`.)
And yeah, I'm fine not making a general helper function, but now I'm just curious - how is this implementation mutating the input (seems it's just reading that input from what I can see)...?

Reply:
Hmm OK, but then we'd have to go over the entire code base - everywhere we've been using test data so far, we've referred to it as mock data. Maybe for consistency leave it as is, and we can go over the entire codebase some time to change that.

As for the mutation: `addAll` calls `thing.add`, which is an RDF/JS Dataset method that mutates the Dataset itself.

Reply:
Just on the mutation thing - sure, it's mutating the instance of the object returned from `createThing()`, but that's a local function variable, not an **_input_** to the function itself. So it seems fine for this function to be treated as a black box (so it can do what it likes internally), but it's not mutating it's input at all, and therefore it's fine, no...?

Reply:
Ehm yep you're right, `mockPolicy` should be safe, `addAll` is not. (Though in a production implementation I'd still prefer it to be built on the primitives we ship ourselves.)

Reply:
Ok, cool (it kills me if I think I'm missing something :) !). On the 'building on our own primitives' question - yeah, I think that very often comes down to a subjective choice. In this particular case though, I actually really like that we are building on a primitive-primitive - i.e. RDF/JS itself is the primitive that our library sits on top of, so in fact I really like that we're proving out that we can use it directly with our `solid-client-js` primitives (arguably we should be proving that out in dedicated tests, rather than it being scattered randomly across our tests, but you get my point!).
It kinda highlights a really good question though - i.e. why did Nic find it **_easier_** to use RDF/JS primitives directly in this test code **_instead_** of `solid-client-js` primitives, which by definition should be 'easier' than RDF/JS....!? (I'm not saying this is highlighting a problem or anything, in fact I really like that we have the choice here, and that our tests are illustrating the ability to exercise that choice, but perhaps our `solid-client-js` SDK is still missing something that Nic could have used here even more naturally...?)

Reply:
@NSeydoux correct me if I'm wrong, but I'm assuming he used the primitives here because that's what we've been doing for the tests for our own primitives, for exactly the reason you mention: to prove that we can use those directly.

However, given that we've already proven that in the unit tests for our primitives, IMO it'd be fine to not do that again in our code that is layered on top of our primitives. (Which is why I asked whether it's OK that I did that in my ACP-related PRs.

Reply:
Exactly as Vincent says: mock data is created using RDFJS primitives not because they are simpler to use, but because we've been using them across the codebase to show interoperability. Also, I feel like using these primitives make the test a tiny bit more DAMP, because we rebuild our primitives "from scratch", so we don't make hidden assumptions on anything our library may add on top of the rdfjs datasets. But I agree that we are now dealing with a higher level of abstraction in the library, layered on top of our own primitives, so it should be fine to use these as well.

Reply:
Ok cool - so we're all the same page then I think (although I'd refer to these examples as 'creating test data', and not 'creating mock data', as the word 'mock' (certainly in my head, and I'm pretty sure across the industry) is associated with mocking **_behaviour_**, and not related to pure data).
So yeah, the choice of using low-level primitives (e.g. RDF/JS constructs), or our higher-level primitives (e.g. Things), will always be a somewhat subjective choice, and I'm fine with either really, but in general I suppose the idea would be to try to limit the number of abstractions to one. So in this particular case, if we agree that ACP is a 'layer on top', then sure, using `solid-client-js` constructs here (as opposed to RDF/JS constructs) would be totally fine.